Support for uncontracted and contracted braille input.

jcsteh opened the pull request:
Beyond the main code in brailleInput to support uncontracted/contracted input, this includes the following changes:
- The list of braille tables has been moved out of the braille module into a separate brailleTables module. Braille tables are now added with a function rather than directly adding them to the data structure. Aside from being necessary in order to specify and check whether a table is contracted, this also makes the data about tables more extensible in future.
- As the data structure for braille tables has now changed and is no longer ordered, this was a good opportunity to sort the list of tables alphabetically when displaying them to the user.
- brailleInput is now notified when reverting config or changing config profiles. This is necessary because brailleInput now maintains some state when the input table is changed.
- brailleInput is now initialised before braille at startup. This is because braille depends on brailleInput to get the currently untranslated input.
- Dot7 and dot8 are now universally bound to braille input specific scripts for erase and enter. Any braille display drivers that had bindings for backspace/enter for braille input have been adjusted accordingly.
- In the User Guide, a "Braille" section has been added above "Application Specific Features". The "Braille Control Types and States" section has been moved to a sub-section of this, and a sub-section on Braille Input has been added.

Fixes #2439. Fixes #6054. Fixes #6113. Fixes #6935.

What's New entries:

In New Features:

```
- You can now type in both contracted and uncontracted braille on a braille display with a braille keyboard. See the Braille Input section of the User Guide for details. (#2439, #6054)
```

In Changes:

```
- The output and input table lists in the Braille Settings dialog are now sorted alphabetically. (#6113)
- Updated liblouis braille translator to 3.2.0. (#6935)
```

jcsteh committed:
Support for uncontracted and contracted braille input.

Beyond the main code in brailleInput to support uncontracted/contracted input, this includes the following changes:
* The list of braille tables has been moved out of the braille module into a separate brailleTables module. Braille tables are now added with a function rather than directly adding them to the data structure. Aside from being necessary in order to specify and check whether a table is contracted, this also makes the data about tables more extensible in future.
* As the data structure for braille tables has now changed and is no longer ordered, this was a good opportunity to sort the list of tables alphabetically when displaying them to the user.
* brailleInput is now notified when reverting config or changing config profiles. This is necessary because brailleInput now maintains some state when the input table is changed.
* brailleInput is now initialised before braille at startup. This is because braille depends on brailleInput to get the currently untranslated input.
* Dot7 and dot8 are now universally bound to braille input specific scripts for erase and enter. Any braille display drivers that had bindings for backspace/enter for braille input have been adjusted accordingly.

jcsteh commented:
@feerrenrut, would you mind reviewing this please? Note that I've tried to explain some of the more obscure changes in the description of this pull request.

feerrenrut commented on source/brailleInput.py:
```diff
@@ -35,18 +41,239 @@ class BrailleInputHandler(object):
+		"""Update the untranslated braille to be shown to the user.
+		If the display will not otherwise be updated, L{updatedisplay} should be called after this.
+		"""
+		self.untranslatedBraille = "".join([unichr(0x2800 + dots) for dots in self.bufferBraille[self.untranslatedStart:]])
```
0x2800? Can this be changed to a named constant?

feerrenrut commented on source/globalCommands.py:
```diff
@@ -1883,6 +1883,17 @@ def script_braille_toFocus(self, gesture):
+
+	def script_braille_enter(self, gesture):
+		brailleInput.handler.enter()
+	script_braille_enter.__doc__= _("Translates any braille input and presses the enter key")
```
missing translators comment

feerrenrut commented on source/gui/settingsDialogs.py:
```diff
@@ -1459,10 +1463,9 @@ def makeSettings(self, settingsSizer):
 		label = wx.StaticText(self, wx.ID_ANY, label=_("&Input table:"))
-		self.inputTableNames = [table[0] for table in braille.INPUT_TABLES]
-		self.inputTableList = wx.Choice(self, wx.ID_ANY, choices=[table[1] for table in braille.INPUT_TABLES])
+		self.inputTableList = wx.Choice(self, wx.ID_ANY, choices=tableChoices)
```
`wx.ID_ANY` is unnecessary, I think this might conflict with `next`

feerrenrut commented on source/speech.py:
```diff
@@ -604,8 +605,21 @@ def speakSelectionChange(oldInfo,newInfo,speakSelected=True,speakUnselected=True
+	@type number: int
+	"""
+	global _suppressSpeakTypedCharactersData
+	_suppressSpeakTypedCharactersData = (number, time.time())
```
Can this be called twice before `number` has had a chance to decrement to 0?

feerrenrut commented on source/speech.py:
```diff
@@ -626,7 +640,21 @@ def speakTypedCharacters(ch):
+		timeOk = time.time() - supTime <= 0.1
+		suppress = number > 0 and timeOk
+		number -= 1
+		if number > 0 and timeOk:
```
I think this logic could be cleaned up so the condition does not need to be repeated. It would have the implication that `_suppressSpeakTypedCharactersData` exists for longer, and number is allowed to reach zero.

eg:
```
if _suppressSpeakTypedCharactersData:
	...
	timeOk = time.time() - supTime <= 0.1

	suppress = number > 0 and timeOk
	if suppress:
		_suppressSpeakTypedCharactersData = (--number, supTime)
	else:
		_suppressSpeakTypedCharactersData = None
else:
	suppress = False
if not suppress and config.conf["keyboard"]["speakTypedCharacters"] and ord(ch)>=32:
	speakSpelling(realChar)
```

feerrenrut commented on source/speech.py:
```diff
@@ -626,7 +640,21 @@ def speakTypedCharacters(ch):
+			_suppressSpeakTypedCharactersData = None
+	else:
+		suppress = False
+	if not suppress and config.conf["keyboard"]["speakTypedCharacters"] and ord(ch)>=32:
```
I'm guessing 32 is a limit for non-printable characters? A named constant would make this more readable.

jcsteh replied:
Quite true. However, as that's very much unrelated code that existed previously (I just added an extra condition), I'd prefer to address this somewhere else.

jcsteh commented on source/brailleDisplayDrivers/brailleNote.py:
```diff
@@ -270,10 +270,10 @@ def display(self, cells):
 			"kb:control+home": ("br(braillenote):space+d1+d2+d3",),
 			"kb:control+end": ("br(braillenote):space+d4+d5+d6",),
-			"kb:enter": ("br(braillenote):space+d8",),
+			"braille_enter": ("br(braillenote):space+d8",),
```
Weird. This doesn't even show up in diffs from command line git. I reckon this is a GitHub bogus thingy. :(

jcsteh commented on source/brailleInput.py:
```diff
@@ -35,18 +41,239 @@ class BrailleInputHandler(object):
+		"""Report a guess about the character(s) produced by a cell of contracted braille.
+		It's not possible to report the exact characters because later cells might change text produced by earlier cells.
+		However, it's helpful for the user to have a rough idea.
+		For example, when typing "alw" for always in English, the user will hear a, l, w.
```
Yes. In English contracted braille, "alw" is the contraction for "always". However, we don't know it will produce "always" until after the word is complete. So, as the user types, they will hear "a l w". I've tried to clarify this comment a bit, but it's a bit tricky to explain.

jcsteh commented on source/brailleInput.py:
```diff
@@ -35,18 +41,239 @@ class BrailleInputHandler(object):
+		"""
+		oldTextLen = len(self.bufferText)
+		oldText = self._translateForReportContractedCell(pos)
+		if oldText[:oldTextLen] != self.bufferText[:oldTextLen]:
```
No reason at all. I'm a moron. :)

jcsteh committed:
Address code review comments.

jcsteh committed:
brailleInput: Fix exceptions when typing a single cell, pressing space, typing two cells then pressing backspace.

This occurred because cellsWithText was being updated even for the space at the end of a word. At this point, we're starting a new word, so cellsWithText needs to be empty.

jcsteh committed:
User Guide: Add note about translation only reflecting the braille word you are typing.

jcsteh committed:
Pressing dot7+dot8 translates immediately but without adding a space or pressing enter.

jcsteh commented:
@feerrenrut, I'm done with this batch of changes. I addressed your code review comments plus some other stuff noted below. Would you mind taking a look again?

@derekriemer reported a bug to me via IRC. STR:

1. Enable input with UEB grade 2.
2. Type "b" and press space.
3. Press dot 6 twice.
4. Press backspace.

This causes an exception. I've fixed this in 0b643d69.

I've also addressed a couple of other issues @derekriemer reported in recent comments on #2439.

jcsteh committed:
Mask braille input in protected fields.

jcsteh commented:
> @feerrenrut, I'm done with this batch of changes.

Actually, I just pushed one more commit to mask characters in protected fields (e.g. password fields). Thanks @derekriemer for catching!

jcsteh committed:
Merge branch 'master' into i2439ContractedBrailleInput

jcsteh committed:
Update to a temporary NV Access branch of liblouis which incorporates several pending upstream changes related to braille input.

jcsteh committed:
Add the Unicode braille input table.

This necessitated the ability to have tables specific to either output or input, rather than assuming all tables can handle both.

jcsteh committed:
Use new liblouis modes to improve back-translation:

* Always set noUndefinedDots so undefined dots (such as indicators that mean nothing by themselves) don't get translated to, for example, "\456/" for dots 4 5 6.
* Set partialTrans when reporting contracted cells, thus avoiding the "zz" hack which was breaking French where z is actually a contraction.

jcsteh commented:
Holding this back from the 2016.4 release, as there are some major outstanding issues:

* [ ] Investigate/fix effect of undefined dots on subsequent translation. For example, with  UEB g2, ⠰⠁⠃ gets translated to "about" instead of "ab" with noUndefinedDots. Without that mode, you get "\56/ab".
* [ ] When there is untranslated input, routing braille outside of the untranslated input should clear the untranslated input.
* [ ] Further UEB back-translation fixes from APH.

jcsteh commented:
Possibly a duplicate of (or at least related to) #6855.
CC @MichaelDCurran.

jcsteh committed:
liblouis sconscript: Add additional source file added in liblouis 3.1.0.

jcsteh committed:
Merge branch 'master' into i2439ContractedBrailleInput

jcsteh committed:
Merge branch 'i6935Liblouis3.1' into i2439ContractedBrailleInput

jcsteh committed:
Update to liblouis 3.1.0, though we're still using a temporary NV Access branch for now due to some changes that aren't yet in 3.1.0.

jcsteh committed:
Merge branch 'master' into i2439ContractedBrailleInput

jcsteh committed:
Upgrade to liblouis 3.2.0.

jcsteh committed:
Don't need liblouis.def any more.

jcsteh committed:
Place indicators at the start and end of braille input so users understand it is not yet part of the actual text.

Re #6956.

jcsteh committed:
Clear braille input state if the caret gets moved (other than by text entry caused by braille input).

jcsteh committed:
Fix keyboardHandler's treatment of Unicode characters sent by SendInput (such as used by brailleInput).

These get sent with a vk code of VK_PACKET and need to be handled specially. Otherwise, we get very strange results because we're trying to treat the scan code as a real keyboard scan code.

jcsteh committed:
braille: Remove RENAMED_TABLES, as this got moved to brailleTables.

jcsteh committed:
brailleInput: Handle renamed tables like we do for braille output, which doesn't apply now but might in future. Don't cause a startup failure if configured input table is invalid.

jcsteh committed:
Switch to a fork of liblouis 3.2.0 which contains some additional back-translation fixes.

jcsteh committed:
readme: Update liblouis version.

josephsl commented:
Hi,

In Liblouis 3.2.0, Greek table name has changed from gr-gr-g1.utb to el.ctb. Thanks.

jcsteh committed:
brailleTables: Update for rename of Greek table in liblouis 3.2.0. Also, remove "grade 1" from the description, since there's no other grade.

jcsteh commented:
@feerrenrut, this is finally ready for another round of review. There have been a lot of changes and quite a long time since you last reviewed it, so you may want to just review from scratch. In case it's helpful, though, the changes since your last review are [5ff07e5b..ff6f627a](https://github.com/nvaccess/nvda/pull/6449/files/5ff07e5b..ff6f627a).

feerrenrut commented on source/brailleTables.py:
```diff
@@ -0,0 +1,338 @@
+#: * fileName: The file name of the table.
+#: * displayname: The name of the table as displayed to the user. This should be translatable.
+#: * contracted: C{True} if the table is contracted, C{False} if uncontracted.
+BrailleTable = collections.namedtuple("BrailleTable", ("fileName", "displayName", "contracted", "output", "input"))
```
This docstring should also have `output` and `input`

feerrenrut commented on source/speech.py:
```diff
@@ -624,7 +642,19 @@ def speakTypedCharacters(ch):
+			_suppressSpeakTypedCharactersTime = None
+	else:
+		suppress = False
+	if not suppress and config.conf["keyboard"]["speakTypedCharacters"] and ord(ch)>=32:
```
Could you replace the value `32` with a name please? I'm not really sure what this is testing for

derekriemer commented on source/brailleTables.py:
```diff
@@ -0,0 +1,338 @@
+#A part of NonVisual Desktop Access (NVDA)
+#This file is covered by the GNU General Public License.
+#See the file COPYING for more details.
+#Copyright (C) 2008-2016 NV Access Limited, Joseph Lee
```
update copyright year

derekriemer commented on source/brailleTables.py:
```diff
@@ -0,0 +1,338 @@
+#brailleTables.py
+#A part of NonVisual Desktop Access (NVDA)
+#This file is covered by the GNU General Public License.
+#See the file COPYING for more details.
```
test test

jcsteh committed:
Switch liblouis to official liblouis master, since all fixes from us have been merged.

jcsteh committed:
Review actions.

jcsteh committed:
BrailleInputGesture: Don't use Python set order for gesture identifiers; we got rid of that in #6962. Make it possible to get the display text from a braille keyboard gesture identifier (as used in the Input Gestures dialog).

jcsteh commented:
@feerrenrut, can you please review the latest commit? Sorry. :(

feerrenrut commented on source/brailleInput.py:
```diff
@@ -378,39 +378,76 @@ class BrailleInputGesture(inputCore.InputGesture):
+		return self._makeDisplayText(self.dots, self.space)
+
+	@classmethod
+	def getDisplayTextForIdentifier(cls, identifier):
```
Where is this called?

feerrenrut replied:
Perhaps a docstring on this to explain what its for? Also, it would be nice to see a couple of examples of what `identifier` might be.

jcsteh committed:
inputCore: Update code doc.

derekriemer commented:
fixes https://github.com/nvaccess/nvda/issues/6956

derekriemer commented:
I'm amazed at how many tickets this autoclosed. What a huge chunk of work.
